drivers/net/nsos_adapt.c

```diff
@@ -0,0 +1,673 @@
+/**
+ * Copyright (c) 2023-2024 Marcin Niestroj
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file
+ *
+ * Linux (bottom) side of NSOS (Native Simulator Offloaded Sockets).
+ */
+
+#include <errno.h>
+#include <fcntl.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <poll.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include "nsos.h"
+#include "nsos_dns.h"
+#include "nsos_errno.h"
+#include "nsos_fcntl.h"
+
+#include "irq_ctrl.h"
+#include "nsi_cpu0_interrupts.h"
+#include "nsi_hws_models_if.h"
+#include "nsi_tracing.h"
+
+#include <stdio.h>
+
+#include <zephyr/sys/dlist.h>
+
+static sys_dlist_t nsos_adapt_sockets = SYS_DLIST_STATIC_INIT(&nsos_adapt_sockets);
```

Comment:
I have reworked the implementation to use 'epoll' instead of 'poll'. That way we don't need to track monitored file descriptors (kernel does that for us) and that made it possible to drop use of linked list.